### ISSUE TITLE
Anti-pattern: Excessive algorithm agility

### ISSUE THREAD

**Issue by msporny** [transmute-industries/lds-jose2020#4]
I'd like to start out with something positive :) --- Great job extending LDS and providing a cryptosuite for JOSE. It's an impressive first cut at doing this and you've gotten a lot of stuff right and have extended LDS in a way that we were not anticipating people would do (which is always nice to see).

Now for the not so great news --- as written, this library greatly expands the attack surface on anyone using it and goes against the whole reason that cryptosuites exist (that is, people that know better should be picking an extremely limited set of options, not enabling a kitchen sink approach, which this library does). 

Unwitting developers are going to pick this library up and have a good chance of blowing their foot off... I haven't tested attacks against this library yet, but it looks like you allow the "none" type to be passed through (which could lead to a variety of nasty attacks). It also looks like you might allows RSA w/ the use of SHA-1? Passing anything the developer wants through to the JOSE library is a *really bad idea* and you shouldn't do it. Linked Data Signatures were designed to not make the same mistakes that JOSE did, one of those things being massive algorithm agility, which leads to very large attack surfaces.

There are currently 15 JWAs that panva/jose library supports... This is the 2020 JOSE cryptosuite... at most, whittle it down to one signature type per key type... pick something that makes sense for most developers for this year... don't let them pick RSASSA-PKCS1-v1_5 when RSASSA-PSS is what they should be using in 2020... ideally, they shouldn't even be using RSA in 2020... Why is ES512 an option? Are we worried that a supercomputer that can consume all of the energy that our sun produces is going to be used to attack a signature used by this library?

Please whittle the choices down and don't make LDS ecosystem suffer the same terrible security choices that plague the JOSE ecosystem.

... and thank you for putting the effort into this library. All of this is intended to be constructive criticism. :)

**Comment by OR13** [comment]
@msporny thanks for this feedback. It was always my intention to show that everything could be supported, and then work with the community to limit the supported JWA and JWK types for this suite. 

We can very easily define a restricted table of supported algorithms for this suite, and throw errors when anything else is detected.

Picking those algorithms is something I'd prefer to tackle in the CCG. 

My hope is submit this as a work item, and then work to restrict the supported algorithms and format to a reasonable set for 2020 as you pointed out, thats not going to be 100% of JWA.

**Comment by OR13** [comment]
If you have a candidate table of key type to alg type mappings (like is in the readme), I will happily take a PR or do the update myself before submitting this as a potential work item.

**Comment by msporny** [comment]
> My hope is submit this as a work item, and then work to restrict the supported algorithms and format to a reasonable set for 2020 as you pointed out, thats not going to be 100% of JWA.

Great, wonderful to hear.

> If you have a candidate table of key type to alg type mappings (like is in the readme), I will happily take a PR or do the update myself before submitting this as a potential work item.

Don't have time for a PR right now (processing issues in the DID spec today)... but here's a first cut that we may want to cut down even further:

* RSASSA-PSS - PS256
* ECDSA - ES256K
* Edwards-curve DSA - EdDSA

That's it for the 2020 suite, everything else should throw an error.

**Comment by OR13** [comment]
https://github.com/transmute-industries/lds-jws2020/pull/7

**Comment by OR13** [comment]
P-256 is desirable to at least Secure Key and probably anyone who wants to use a hardware KMS, or non extractable web keys.

**Comment by kdenhartog** [comment]
In New Zealand, we've got requirements to support P-384 using ECDSA. We'd like to see support for this as well. Additionally, I'm looking for some reference materials to decide on if we should be using probabilistic signature schemes vs deterministic signature schemes. Essentially should we be using PS(256/384/521) or ES(256/384/521). Ideally, we narrow it to one based on expert recommendations and select the same keysize, but we're willing to support others that are required with the same goal in mind of not adding the kitchen sink in.

**Comment by OR13** [comment]
we might as well add support for all the NIST curves, if we are gonna support P-256, this is trivial to do now.

**Comment by msporny** [comment]
> In New Zealand, we've got requirements to support P-384 using ECDSA.

Citation please. I believe you, just trying to see if we need to upgrade everyone to P-384.

> we might as well add support for all the NIST curves

I don't think you actually mean that. There are at least 25 NIST curves in 186-4, IIRC including prime, binary, pseudorandom, koblitz, etc... so, huge -1 to "support for all the NIST curves".

If you did mean to imply that you're opening up support for 25 curve types, then we're talking past each other.

One safe choice for 2020 per key type ... get it down to one safe choice for each curve and, in general, only support one curve for each signature type.

This discussion is also the exact reason we decided to not do a multivariate signature cryptosuite... it's exceedingly difficult to pick something that the whole world can agree on. The differences between New Zealand (P-384) and the US (P-256) is one example.... and that's between two countries that trust each other! :)

So, when is this suite going to support China's SM9-IBS or Russia's GOST R 34.10-2012? :)

**Comment by OR13** [comment]
> So, when is this suite going to support China's SM9-IBS or Russia's GOST R 34.10-2012? :)

As soon as those are here:

https://www.iana.org/assignments/jose/jose.xhtml

The counter argument is that library authors are not the ones who should be making these kinds of choices, and instead we should be following IANA: Required, Recommended, Recommended+, Optional

Just like https://www.npmjs.com/package/jose

I'd rather not be political about this, and just say "lds-jws2020" supports all Required, Recommended, Recommended+, Optional JWS.... but we are here debating a smaller list because politics is a universal language :)

From an implementation burden perspective, we generate the supported alg documentation from a single source: https://github.com/transmute-industries/lds-jws2020/blob/master/src/__tests__/__fixtures__/supportedKeyTypes.js

All you need to do is add your key type to that file, and update this mapping here:

https://github.com/transmute-industries/lds-jws2020/blob/master/src/getRecomendedAlg.js

And support will be added.

There is nothing other than politics stopping us from supporting everything and making recommendations instead of choosing specifically to support less algorithms then npm jose.

**Comment by kdenhartog** [comment]
> Citation please. I believe you, just trying to see if we need to upgrade everyone to P-384.

https://www.nzism.gcsb.govt.nz/ism-document/#2106

Of note, NSA also recommends the use of P-384 as well for up to top secret documents. You’ll have to hit the get file button to view the document. I don’t think I can link directly to it.

https://apps.nsa.gov/iaarchive/library/ia-guidance/ia-solutions-for-classified/algorithm-guidance/commercial-national-security-algorithm-suite-factsheet.cfm

**Comment by selfissued** [comment]
For what it’s worth, I’m actually fine with the DID WG profiling the JOSE algorithms, choosing its own set of REQUIRED, RECOMMENDED, etc. algorithms.  There have been a lot of developments since the JOSE specs were finalized and our recommendations can reflect what we consider to be current best practices for our use cases, if we choose.

**Comment by kdenhartog** [comment]
I was thinking we could register an additional optional header like `profile` and then use the `crit` header to indicate it must be processed, which would be validating that algorithms indicated match the profile. This would add profiles, but not break processing of the cryptography in JOSE implementations today.

Thoughts on this approach?

**Comment by OR13** [comment]
@kdenhartog can you provide a json example of what you mean?

**Comment by kdenhartog** [comment]
```
{
  header: {
    "enc": "XC20P",
    "alg": "ECDH-ES+XC20PKW",
    "profile": "BernsteinSuite2020",
    "crit": ["profile"]
  },
  Recipients: [{
     { "encrypted_key":"XCN8V57Q1tgauLW1zgj1WQZQ0WDV0QwyEZFNNwF6DmiI49CK5s58dsV0d_NZK-SMNqe0iFXgXFvHpoc8kuViSM_K5lrpbMStj7CRPskvbS8=",
            "header": {
                "epk": {
                     "kty":"OKP","crv":"Ed25519",
                     "x":"11qYAYKxCrfVS_7TyWQHOg7hcvPapiMlrwIaaPcHURo"
                }
            }
}
```

Where BernsteinSuite2020 indicates ECDH-ES+XC20PKW, XC20P, and Ed25519 must be used with the ECDH scheme.

**Comment by mavarley** [comment]
I just wanted to add my support to the position that production crypto-suites should be limited to the accepted algorithms for the deployment/governance/regulatory framework - and I think that LD can aid in any scenarios where layering/crossover is needed (@msporny @OR13 correct me if I am wrong?) [ "NISTP-P256" ], ["NZA-P384" ] combined [ "NIST-P256", "NIST-P384" ] ?

Although developers may find a crypto-suite "AllJOSE" useful in development as prototypes and proof of concepts get iterated, any production system will need to be locked down to a certain set of approved algorithms. So maybe if an "AllJOSE" definition exists, it should be called "AllJOSEDevOnly" :) (and yes, I can see arguments against even this definition, as dev code may 'escape' to production despite best preventative efforts ).

**Comment by msporny** [comment]
> LD can aid in any scenarios where layering/crossover is needed (@msporny @OR13 correct me if I am wrong?) [ "NISTP-P256" ], ["NZA-P384" ] combined [ "NIST-P256", "NIST-P384" ] ?

I'd go further and state that we should force the decision between NIST-P384 and NIST-P256 in a LD cryptosuite. That is, For JwsSignature2020, you can only choose NIST-P384 (for the secp256r1 curve)... full stop.

**Comment by msporny** [comment]
> I'm realizing now what I'm suggesting is a bit orthogonal to the purpose of the library though. 

Yeah, this lib is for signatures, not encryption... there used to be a Linked Data Encryption spec a very, very long time ago, but we dropped it because it didn't really do anything extra that you couldn't already do w/ JWE and CWE. Note that the Encrypted Data Vault spec and implementations that Digital Bazaar has done just uses JWE and JSON (no JSON-LD or Linked Data anything at the base protocol... although, you can put JSON-LD in the encrypted content).

**Comment by kdenhartog** [comment]
Looking at JOSE, I think this could still be useful for specifying things with EdDSA. For some reason the spec was really specific with the inital algorithms ES256( ECDSA + P-256 using SHA-256), but EdDSA can be used with both Ed25519 and Ed448 which don't have individual algorithm values marked in the IANA registry.

Additionally, if we want to support other signature schemes (I realize the whole point of this is to avoid this but bare with me for a second) then this would be useful. For example, if I wanted to support BLS signatures using Ed25519 that could be a separate suite. This would be useful for having a scheme for threshold signatures, but we wouldn't do one for multiple curves (e.g. not supporting BLS sigs with Ed25519, Ed448, and BLS12-381)

**Comment by OR13** [comment]
The original goal of this suite, was to support all of jose JWS, and let developers choose, just like they can with jose today.

I'm increasingly leaning towards doing that, and putting clear warnings at the top, saying that just like with JOSE, you have a responsibility to choose carefully what you will support, from what is available.

**Comment by msporny** [comment]
> The original goal of this suite, was to support all of jose JWS, and let developers choose, just like they can with jose today.

-1 

The title of this issue is: "Anti-pattern: Excessive algorithm agility"... letting developers choose is an anti-pattern that harms the industry. It has led to a very long list of preventable security failures that have led to hundreds of millions of dollars worth of damage. Please do not do that. If someone wants to shoot themselves in the foot, have them build the crossbow that does that... don't provide one to them.

**Comment by kdenhartog** [comment]
I'd rather we choose as a community what we want to support rather than move in the direction of supporting everything. With that in mind, we should be having that discussion soon before too many other curves and algorithms get introduced.

**Comment by mavarley** [comment]
OK - I believe I misunderstood the goal of this library, maybe. (I thought we were building the crossbow).

I do think defining a subset of 'safe' and 'community interoperable' JOSE algorithms is very valuable - and maybe this spec should morph into that? Or do we need to start fresh?

(@msporny I believe you are suggesting we morph this spec into the 'approved' subset, so the CrossbowAllJOSE suite never escapes the lab - because it doesn't exist?)

**Comment by OR13** [comment]
Shouldn't the spec of a restricted suite of JWS algs have a name like "Restricted/W3CFavJWS2020", or similar?

It was my original intention to save myself, and other developers from needing to make more Linked Data Signature Suites... specifically suites for P-256 and RSA respectively.

I have had a number of conversations regarding hardware support / government regulatory requirements, where some of the "less desirable" crypto are required / only thing that works.

I already have called out the use of specific limited LDS suites like Ed25519Signature2018, for cases where you know you only need to support that... and pointed to safe curves. 

This issue could have been opened just as easily on the jose library repo, the intention of this repo was to support LDS for that library.

**Comment by panva** [comment]
> This issue could have been opened just as easily on the jose library repo

Why would it? The developers are encouraged to provide a set of algorithms they only want to accept on the verify function call if they’re working with keys fetched from the wild.

**Comment by OR13** [comment]
developers are able to do that by only choosing to use Ed25519Signature2018, and by choosing this library or a restricted subset of this library they would be getting a set / subset of jose.

Seems to me that this is best solved by just exposing jose jws options as is for JWKs, and then publishing forked version of this library that support less of jws... or waiting for someone else to do that.

@panva do you publish a subset of jose with recommended algs only? because that seems to be what people are suggesting I do here... what do you think I should do?

**Comment by panva** [comment]
@OR13 no I don't, I follow the IANA registry and don't implement deprecated or prohibited algorithms. The moment any WG or spec updates the registry to change an alg's `JOSE Implementation Requirements` registry field i'd release a new major version.

> Seems to me that this is best solved by just exposing jose jws options as is for JWKs, and then publishing forked version of this library that support less of jws... or waiting for someone else to do that.

Why not just add a profile option that limits the set of algs/keys? You can add easily add multiple profiles based on different country/regulatory/spec requirements.

**Comment by OR13** [comment]
Agreed regarding forking, thats what I'm trying to avoid... Interesting, thats essentially what @kdenhartog suggested above, we could do:

```
{
  header: {
    "alg": "EdDSA",
    "profile": "BernsteinSuite2020",
    "b64": false,
    "crit": ["profile", "b64"]
  },
...
}
```

Then as preferences change (next week / year), consumers would just need change the profiles they want to support... I'm not very familiar with using profiles outside of the OAuth JWT use case...

If we define `BernsteinSuite2020` somewhere in w3c / IETF and let consumers of this library choose does this solve this issue?

**Comment by panva** [comment]
I’m not saying to define a new crit field. Altho its a more strict option I have yet to see a library other than mine that uses crit properly.

**Comment by OR13** [comment]
My first exposure to `crit` was here: https://github.com/digitalbazaar/jsonld-signatures/blob/fba9caf5b3c005b5dbae461801986761295bbc12/lib/suites/JwsLinkedDataSignature.js#L75

Per your docs: https://github.com/panva/jose/blob/HEAD/docs/README.md#jwsverifyjws-keyorstore-options

Any library consuming this suite would just pass a list of supported algs when initializing the suite.

@msporny anyone using vc-js could choose to limit the supported algs using this approach, does this sound like less of a foot crossbow? I don't want to have to keep making linked data signature suites for jws every-time the wind changes... even if it changes every few years.

**Comment by msporny** [comment]
@mavarley wrote:
> (@msporny I believe you are suggesting we morph this spec into the 'approved' subset, so the CrossbowAllJOSE suite never escapes the lab - because it doesn't exist?)

Yes, correct.

> Shouldn't the spec of a restricted suite of JWS algs have a name like "Restricted/W3CFavJWS2020", or similar?

No, it should be JwsSignature2020 and pick the limited subset of options that make sense if you're writing software in 2020.

> This issue could have been opened just as easily on the jose library repo, the intention of this repo was to support LDS for that library.

No, because the JOSE community seems to be totally fine with providing foot crossbows as evidenced by the long list of foot crossbow JWS libraries that exist out there... 

> If we decided to support all, I think its very reasonable for anyone to fork and support a subset only.

Don't do that, it'll be painful for you, as @panva has pointed out.

> @msporny anyone using vc-js could choose to limit the supported algs using this approach, does this sound like less of a foot crossbow?

No, it's still a foot crossbow.... because the vast majority of developers don't know how to limit the choices, so they just turn everything on and go. Again, this is what we see time and time again when there are security failures at organizations. Some newbie developer that wasn't properly trained was asked to secure the site... so they pick a library off of the shelf, which has all the options turned on (because who wants to deal w/ developers that complain that your library doesn't work for them)... and then they get it working and deploy the software into production and don't look back.

You are writing a security library - don't make it fail open. Security libraries should fail closed.

https://www.ixiacom.com/company/blog/fail-closed-fail-open-fail-safe-and-failover-abcs-network-visibility

>  I don't want to have to keep making linked data signature suites for jws every-time the wind changes... even if it changes every few years.

Then you are missing one of the fundamental design goals of Linked Data Security. We have decades of examples of this *not* being done and the security breaches that we have seen over that same time frame is a testament for why designing systems in this way is a really bad idea.

At the W3C DID WG F2F, @ChristopherA outlined that one of the mistakes that was made in the early SSL/TLS was not having an expiration date on cryptographic suites... which is why people are *still* using cryptography that is known to be compromised during downgrade attacks.

Security characteristics of these libraries *do* change every couple of years and they *should* be updated every couple of years. Security is hard... you can't just implement something and walk away from it. It will eventually fail... and when it fails, you don't want it to fail open (which is the really dangerous thing that is being proposed in this thread).

Make the hard choices, protect the people that use your cryptosuite and library. Almost all of them do not have the knowledge to make the right choice.

**Comment by troyronda** [comment]
I'm not understanding the rationale of "forcing" NIST P-384 vs NIST P-256 in this case. There are always tradeoffs (e.g., performance, implementation availability, common usage, ...). 

What is the required security level that this "forcing" function is being based upon? Do the other curves listed meet that same bar?
